### ISSUE TITLE
"No symbols found" after "Go to symbol in file" in C/C++

### ISSUE THREAD

**Issue by Trucoto** [microsoft/vscode-cpptools#1126]
* Operating System and version: Ubuntu 17.04
* VS Code version: 1.17.1
* C/C++ extension version: 0.13.1
* step-by-step instructions to reproduce the issue:

1) Press Ctrl-Shift-O
2) After the @ there is nothing and under it "No symbols found". No busy indication (no bar coming and going as if looking for symbols)

This used to work. I guess it stopped working after last upgrade of C/C++ extension to 0.13.1

* Relevant settings from your settings.json, c_cpp_properties.json, and/or launch.json files:

No "browse" section, no "databaseFilename" entry. "intelliSenseMode" is "clang-x64". "search.useRipgrep" is false and "C_Cpp.intelliSenseEngine" is "Default".

* Other tests: I removed everything ~/.config/Code/User/workspaceStorage; databases were recreated but to no avail.

**Comment by bobbrow** [comment]
If your "browse" section is missing, then the extension doesn't know where to look for files.  Specifically, the "path" property in the "browse" section tells the extension which folders contain the code you care to index.  Recently we updated the extension to clean up the symbol database to conform to the "browse.path" setting since users were seeing out-of-date recommendations.

If you add the paths back to your "browse" section, the extension will re-index your code and this should start working again.

**Comment by Trucoto** [comment]
I am not sure I understand the settings. I have on the one hand "includePath" with basically the -I list of directories where include files are supposed to be. The help provided by vscode is 

> A list of paths for the IntelliSense engine to use while searching for included headers. Searching on these paths is not recursive. 

I am using the new engine and not the tag parser. On the other hand, browse.path help tells me 

> A list of paths for the tag parser to use while searching for included headers. Searching on these paths is recursive by default. Specify '\*' to indicate non-recursive search. For example: '/usr/include' will search through all subdirectories while '/usr/include/*' will not. 


So which is used for what when I use the new engine and not the tag parser? Should I use both? With the same list of directories? Without "browse.path", Intellisense has no problem on resolving includes, by the way. I can hover over symbols and see correct information on them. My only problem is that  "Go to symbol"  stopped working. "Go to definition" and "Go to declaration" are not working either, though vscode seems to know perfectly what kind of symbol is, how it was declared, etc.

**Comment by Trucoto** [comment]
I added 

```
			"browse": {
				"path": [
					"${workspaceRoot}"
				]
			}
```

And it never stops parsing, i.e., the fire icon is always on, "Go to symbol" has a bar always moving indicating a wait state, and hover always says "Loading...". It's been an hour already and it's a fairy small project.

**Comment by sean-mcmanus** [comment]
Go to Symbol and Go to Declaration only uses the browse.path setting (an implementation using the includePath settings hasn't been implemented yet). If hover says Loading, it means your main thread is stuck for some reason (i.e. a bug). Can you try using a subfolder of the workspaceRoot with less files? You could also try adding `"limitSymbolsToIncludedHeaders": true` to the browse settings (or deleting your c_cpp_properties.json and re-creating it. It's possible we could have a bug in scenarios where expected settings are missing. Enabling logging might also help diagnose the problem: https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/Enabling%20logging.md .

**Comment by hellspwj** [comment]
I have the same problem.

**Comment by Trucoto** [comment]
I followed all suggestions (recreated c_cpp_properties.json, limitSymbolsToIncludedHeaders: true, a smaller source dir), but to no avail. I have some more information, though. I have the same project (same dir) that is compiled under two different toolchains, but they share most of the source files. Here is the complete configuration for both:

```
{
    "configurations": [
        {
            "name": "Linux x86_64",
            "includePath": [
                "/usr/include/c++/6",
                "/usr/include/x86_64-linux-gnu/c++/6",
                "/usr/lib/gcc/x86_64-linux-gnu/6/include/",
                "/usr/local/include",
                "/usr/lib/clang/4.0.0/include",
                "/usr/include",
                "/usr/include/x86_64-linux-gnu",
                "${workspaceRoot}",
                "${workspaceRoot}/common",
                "${workspaceRoot}/Linux",
                "${workspaceRoot}/Linux/x86_64",
                "${workspaceRoot}/Linux/x86_64/openssl/include",
                "${workspaceRoot}/Linux/x86_64/zlib/include"
            ],
            "defines": [
                "P_64",
                "P_LINUX",
                "MACHINE=x86_64"
            ],
			"intelliSenseMode": "clang-x64",
			"browse": {
				"path": [
					"${workspaceRoot}"
				],
				"limitSymbolsToIncludedHeaders": true,
				"databaseFilename": ""
			}

        },
        {
            "name": "uClinux arm7",
            "includePath": [
                "/root/bin/arm-uclinux-tool/lib/gcc/arm-uclinux/3.4.0/include",
                "/root/bin/arm-uclinux-tool/arm-uclinux/include",
                "/root/bin/arm-uclinux-tool/include/c++/3.4.0/",
                "/root/bin/arm-uclinux-tool/include/c++/3.4.0/arm-uclinux/",
                "${workspaceRoot}",
                "${workspaceRoot}/common",
                "${workspaceRoot}/uClinux",
                "${workspaceRoot}/uClinux/arm7",
                "${workspaceRoot}/uClinux/arm7/openssl/include",
				"${workspaceRoot}/uClinux/arm7/zlib/include",
				"${workspaceRoot}/uClinux/vic"
            ],
            "defines": [
                "P_LINUX",
                "P_ARM7",
                "P_UCLINUX",
                "P_32"
			],
			"intelliSenseMode": "clang-x64",
			"browse": {
				"path": [
					"${workspaceRoot}"
				],
				"limitSymbolsToIncludedHeaders": true,
				"databaseFilename": ""
			}
        }
    ],
    "version": 3
}
```
Now, for a clean test, I removed all logs and all databases. When I open the source dir, it starts by default with the "uClinux arm7" configuration. After some time parsing, the fire icon goes off; however I have "No symbols found" under "Go to symbol". I can hover symbols and see information, but "Go to definition" and "Go to declaration" are not working. Then I switch to the other profile, "Linux x86_64". The fire icon is on, and never goes out. I had the environment variables on for logs with verbosity = 5. If I exit at this point, there are two logs. One is empty, the other one only says:

```
Log initialized. Verbosity level = 5
handle_compile
handle_load
handle_get_errors
updating translation unit, range [0, 64187], preparse [0]
errorSquiggles count: 0

```

Now, if I repeat the test from the start (removing logs, databases, etc.) but I exit before switching to the "Linux x86_64" profile, I have information on the logs. One starts like this:

```
Log initialized. Verbosity level = 5
initialize
using the new IntelliSense engine
IntelliSense engine fallback is ON
autocomplete is enabled
error squiggles are enabled
File exclude (unconditional): **/.git
File exclude (unconditional): **/.svn
File exclude (unconditional): **/.hg
File exclude (unconditional): **/CVS
File exclude (unconditional): **/.DS_Store
File exclude (unconditional): **/*.o
File exclude (unconditional): **/*.d
File exclude (unconditional): **/.vscode
Search exclude (unconditional): **/node_modules
Search exclude (unconditional): **/bower_components
Search exclude (unconditional): **/*.o
Search exclude (unconditional): **/*.d
Search exclude (unconditional): **/.vscode
workspace/didChangeConfiguration
using the new IntelliSense engine
IntelliSense engine fallback is ON
autocomplete is enabled
error squiggles are enabled
File exclude (unconditional): **/.git
File exclude (unconditional): **/.svn
File exclude (unconditional): **/.hg
File exclude (unconditional): **/CVS
File exclude (unconditional): **/.DS_Store
File exclude (unconditional): **/*.o
File exclude (unconditional): **/*.d
File exclude (unconditional): **/.vscode
Search exclude (unconditional): **/node_modules
Search exclude (unconditional): **/bower_components
Search exclude (unconditional): **/*.o
Search exclude (unconditional): **/*.d
Search exclude (unconditional): **/.vscode
textDocument/didOpen
cpptools/queryDefaultSdks
cpptools/didChangeFolderSettings
add folder to browse.path: /home/leandro/Desarrollo/hsl/working
message_handler::parse_root()
    processing folder (recursive): /home/leandro/Desarrollo/hsl/working/

```
and follows with a number of "tag_parse failed 0". After that, it says:

```
generate_file_list: 1600 file(s) processed
Database safe to open

```
And ends with termination information. The other one just says:

```
Log initialized. Verbosity level = 5
handle_compile
handle_load
handle_get_errors
updating translation unit, range [0, 64187], preparse [0]
errorSquiggles count: 0
handle_quick_info
handle_quick_info
handle_quick_info

```

Now I remove all logs again but not the databases, start vscode, which again begins with the default profile. Then I do the switch to "Linux x86_64", there's a brief fire but this time it stops. Still "no symbols found" on both profiles, but it does not hang parsing. At exit I have the following log:

```
Log initialized. Verbosity level = 5
initialize
using the new IntelliSense engine
IntelliSense engine fallback is ON
autocomplete is enabled
error squiggles are enabled
File exclude (unconditional): **/.git
File exclude (unconditional): **/.svn
File exclude (unconditional): **/.hg
File exclude (unconditional): **/CVS
File exclude (unconditional): **/.DS_Store
File exclude (unconditional): **/*.o
File exclude (unconditional): **/*.d
File exclude (unconditional): **/.vscode
Search exclude (unconditional): **/node_modules
Search exclude (unconditional): **/bower_components
Search exclude (unconditional): **/*.o
Search exclude (unconditional): **/*.d
Search exclude (unconditional): **/.vscode
workspace/didChangeConfiguration
using the new IntelliSense engine
IntelliSense engine fallback is ON
autocomplete is enabled
error squiggles are enabled
File exclude (unconditional): **/.git
File exclude (unconditional): **/.svn
File exclude (unconditional): **/.hg
File exclude (unconditional): **/CVS
File exclude (unconditional): **/.DS_Store
File exclude (unconditional): **/*.o
File exclude (unconditional): **/*.d
File exclude (unconditional): **/.vscode
Search exclude (unconditional): **/node_modules
Search exclude (unconditional): **/bower_components
Search exclude (unconditional): **/*.o
Search exclude (unconditional): **/*.d
Search exclude (unconditional): **/.vscode
textDocument/didOpen
cpptools/queryDefaultSdks
cpptools/didChangeFolderSettings
add folder to browse.path: /home/leandro/Desarrollo/hsl/working
message_handler::parse_root()
    textDocument/didOpen
    processing folder (recursive): /home/leandro/Desarrollo/hsl/working/
        checking for syntax errors: file:///home/leandro/Desarrollo/hsl/working/common/crypto_openssl.cc
        queue_get_errors_for_file: file:///home/leandro/Desarrollo/hsl/working/common/crypto_openssl.cc
                        textDocument/codeAction
            errorSquiggles count: 0
    generate_file_list: 1600 file(s) processed
    generate_file_list: 0 file(s) removed from database
    message_handler::parse_root() -- done searching for files.
    processing folder (recursive): /home/leandro/Desarrollo/hsl/working/
    generate_file_list: 1600 file(s) processed
cpptools/didChangeSelectedSetting
shutting down IntelliSense server: /home/leandro/Desarrollo/hsl/working/common/crypto_openssl.cc
shutdown request sent and acknowledged by server
exiting _receive_comm
shutting down IntelliSense server: /home/leandro/Desarrollo/hsl/working/common/crypto_openssl.cc
checking for syntax errors: file:///home/leandro/Desarrollo/hsl/working/common/crypto_openssl.cc
add folder to browse.path: /home/leandro/Desarrollo/hsl/working
message_handler::parse_root()
    queue_get_errors_for_file: file:///home/leandro/Desarrollo/hsl/working/common/crypto_openssl.cc
    processing folder (recursive): /home/leandro/Desarrollo/hsl/working/
            errorSquiggles count: 1
            textDocument/codeAction
    generate_file_list: 1600 file(s) processed
    generate_file_list: 0 file(s) removed from database
    message_handler::parse_root() -- done searching for files.
    processing folder (recursive): /home/leandro/Desarrollo/hsl/working/
            textDocument/hover
            textDocument_hover(): [672, 49]
    generate_file_list: 1600 file(s) processed
Database safe to open
textDocument/documentSymbol
shutdown
shutting down IntelliSense server: /home/leandro/Desarrollo/hsl/working/common/crypto_openssl.cc
shutdown request sent and acknowledged by server
exiting _receive_comm
Database safe to open
shutting down IntelliSense server: /home/leandro/Desarrollo/hsl/working/common/crypto_openssl.cc
exit complete

```
Those steps produce exactly the same behavior if I repeat them, i.e., a clean start with no databases produces no symbols on the first profile and parses forever if I switch to the second one without exiting first. If I exit instead of switching right in the clean run, I can switch the second time I run vscode without producing a neverending parsing. On both cases there are no symbols, though.

**Comment by bobbrow** [comment]
Just a shot in the dark, but do you happen to have a "Microsoft.VSCod" zombie process showing in your Activity Monitor?  I'm seeing some weird behavior in our process forking that shows up from time to time.  The forked child process is supposed to be swapped out by execv, but it gets stuck and holds on to file descriptors and interferes with all communication to VS Code making the extension completely unresponsive.

Reloading the window releases it and allows for symbols to appear again, but I'm still trying to figure out the root cause for this.  I'm working on some other linux process issues right now that I hope to have addressed before the next release.

**Comment by Trucoto** [comment]
Right, I had several zombie processes called **Microsoft.VSCode.CPP.Extension.linux**. Killing them (some of them had to be with the 9 signal) and restarting vscode brought the "Go To Symbol" back to normal for the default configuration on fresh start. Switching to the other configuration ("Linux x86_64") hangs parsing again, however, and leaves a new zombie. It's completely reproducible with the above steps for me.

**Comment by wming126** [comment]
I have the same problem. 
Thanks!

**Comment by SkyKoo** [comment]
Yes, I have the same problem too.
Thanks!

**Comment by Trucoto** [comment]
Any update on this issue, @bobbrow? It is marked as "more info needed" when I supplied all info requested

**Comment by sean-mcmanus** [comment]
Your configs are sharing the same database -- that will cause problems. You should set the databaseFilename to something unique to each configuration. Does that fix it? It sounds like the problems start after switching configurations.

**Comment by Trucoto** [comment]
I added a different database filename for each configuration. Now I have:

```
{
	"configurations":
	[
        {
            "name": "Linux x86_64",
            "includePath": [
				"/usr/lib/gcc/x86_64-linux-gnu/7/include",
                "/usr/local/include",
                "/usr/lib/gcc/x86_64-linux-gnu/7/include-fixed",
                "/usr/include/x86_64-linux-gnu",
				"/usr/include",
				"/usr/include/c++/7",
				"/usr/include/x86_64-linux-gnu/c++/7",
				"/usr/include/c++/7/backward",
                "${workspaceRoot}",
                "${workspaceRoot}/common",
                "${workspaceRoot}/Linux",
                "${workspaceRoot}/Linux/x86_64",
                "${workspaceRoot}/Linux/x86_64/openssl/include",
                "${workspaceRoot}/Linux/x86_64/zlib/include"
            ],
			"defines":
			[
                "P_64",
                "P_LINUX",
                "MACHINE=x86_64"
            ],
			"intelliSenseMode": "clang-x64",
			"browse":
			{
				"path": [ "${workspaceRoot}" ],
				"limitSymbolsToIncludedHeaders": true,
				"databaseFilename": "$[workspaceRoot}/.vscode/linux.db"
			}
        },
        {
            "name": "uClinux arm7",
			"includePath":
			[
                "/root/bin/arm-uclinux-tool/lib/gcc/arm-uclinux/3.4.0/include",
                "/root/bin/arm-uclinux-tool/arm-uclinux/include",
                "/root/bin/arm-uclinux-tool/include/c++/3.4.0/",
                "/root/bin/arm-uclinux-tool/include/c++/3.4.0/arm-uclinux/",
                "${workspaceRoot}",
                "${workspaceRoot}/common",
                "${workspaceRoot}/uClinux",
                "${workspaceRoot}/uClinux/arm7",
                "${workspaceRoot}/uClinux/arm7/openssl/include",
				"${workspaceRoot}/uClinux/arm7/zlib/include",
				"${workspaceRoot}/uClinux/vic"
            ],
			"defines":
			[
                "P_LINUX",
                "P_ARM7",
                "P_UCLINUX",
                "P_32"
			],
			"intelliSenseMode": "clang-x64",
			"browse":
			{
				"path": [ "${workspaceRoot}" ],
				"limitSymbolsToIncludedHeaders": true,
				"databaseFilename": "$[workspaceRoot}/.vscode/uclinux.db"
			}
        }
    ],
    "version": 3
}
```

The behavior does not change. These are the steps:

1) Change configuration file. 
2) Close vscode
3) **rm -rf ~/.config/Code/User/workspaceStorage/***
4) On my working directory, **ls -a .vscode** (that returns only **c_cpp_properties.json**  and **settings.json**)
5) Start vscode
6) Browse around. The fire icon is on for a little while. "Go to symbol" still not working.
7) Close vscode.
8) **ls -l  ~/.config/Code/User/workspaceStorage/*** shows a newly created folder **ms-vscode.cpptools**.
9) **ls -a .vscode** on my working directory still returns the same two configuration files and no new database.

What am I doing wrong?

**Comment by sean-mcmanus** [comment]
The ms-vscode.cpptools folder seems okay, but is it empty or does it have a database file in it? The fact that no database is created at "${workspaceRoot}/.vscode" appears to be a problem.

**Comment by Trucoto** [comment]
Yes, there is a .browse.VC.db in ms-vscode.cpptools.

**Comment by sean-mcmanus** [comment]
Your databaseFilename has the incorrect `$[workspaceRoot}`, which should be `${workspaceRoot}`. It's causing the default to still be used because the path isn't resolved correctly. Yeah, we should be showing more errors.

**Comment by Trucoto** [comment]
Yes! Sorry for that. I corrected it. As a result, I have now symbols after "Go to symbol in file", but only on certain source files, while in others I have "No symbol found" as before. On a small project, 32 source files worked while 15 didn't.

**Comment by sean-mcmanus** [comment]
It sounds like something in those files is causing our symbol parser to fail. One thing to look for is #define macros that if undefined can mess up parsing, such as attributes that go after a function declaration, which need be added to the cpp.hint file in our extension root directory. Also, if you turn the C_Cpp.loggingLevel to "Warning", you might see a warning about "Tag parsing failed: " in the C/C++ output panel. The extension process could also be crashing during parsing. We also don't give symbols for local variables. I'm not able to repro any problem so far, so any additional info would help (your library files could be causing a problem).

**Comment by sean-mcmanus** [comment]
This may be fixed in our pre-release if you want to try it out (https://github.com/Microsoft/vscode-cpptools/releases/tag/v0.14.6-insiders) . There was at least one bug fixed that could have caused this. You may need to do a Reset IntelliSense Database command though.

**Comment by hellspwj** [comment]
it can work normally

**Comment by sean-mcmanus** [comment]
We think we fixed this issue with 0.14.6, although if your code has UTF-8 characters > 1 byte in it, we may still fail (there's another issue on that).

**Comment by littlebirdxu** [comment]
I also have encountered the same problem that is because of i work on a remote machine which is linux. But my local machine is a Mac-book. i use ssh for the communication. At last, i resolve this problem by install the extension: c/c++. Use the VISX from: [https://github.com/Microsoft/vscode-cpptools/releases](https://github.com/Microsoft/vscode-cpptools/releases), and choose the right linux version for my remote machine. Hope this is helpful~
